File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        return self.divide(lists, 0, len(lists) - 1)
    
    def divide(self, arr, l, r):
        if l > r:
            return None
        if l == r:
            return arr[l]
        
        m = l + (r - l) // 2 
        left = self.divide(arr, l, m)
        right = self.divide(arr, m + 1, r)

        return self.merge(left, right)


    def merge(self, l1, l2):
        res = ListNode()
        curr = res

        while l1 and l2:
            if l1.val <= l2.val:
                curr.next = l1
                l1 = l1.next
            else:
                curr.next = l2
                l2 = l2.next
            curr = curr.next
        
        if l1:
            curr.next = l1
        else:
            curr.next = l2
        
        return res.next

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-2.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} k
     * @return {ListNode}
     */
    reverseKGroup(head, k) {
        let dummy = new ListNode(0, head);
        let groupBefore = dummy;

        while (true) {
            let groupEnd = groupBefore;
            for (let i = 0; i < k; ++i) {
                groupEnd = groupEnd.next;
                if (!groupEnd) {
                    break;
                }
            }

            if (!groupEnd) {
                break;
            }

            let groupAfter = groupEnd.next;
            let groupStart = groupBefore.next;

            let prev = groupAfter;
            let curr = groupStart;

            while (curr != groupAfter) {
                let tmp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = tmp;
            }

            groupBefore.next = groupEnd;
            groupBefore = groupStart;
        }

        return dummy.next;
    }
}
